compiler-core/src/type_/pattern.rs

```diff
@@ -75,6 +75,47 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
         }
     }
 
+    fn insert_constant(
+        &mut self,
+        name: &str,
+        literal: Constant<Arc<Type>, SmolStr>,
+        location: SrcSpan,
+    ) -> Result<(), UnifyError> {
+        match &mut self.mode {
+            PatternMode::Initial => {
+                // Register usage for the unused variable detection
+                self.environment
+                    .init_usage(name.into(), EntityKind::PrivateConstant, location);
+                // Ensure there are no duplicate constant names in the pattern
+                if self.initial_pattern_vars.contains(name) {
+                    return Err(UnifyError::DuplicateVarInPattern { name: name.into() });
+                }
+                // Record that this variable originated in this pattern so any
+                // following alternative patterns can be checked to ensure they
+                // have the same variables.
+                let _ = self.initial_pattern_vars.insert(name.into());
+                // And now insert the variable for use in the code that comes
+                // after the pattern.
+                self.environment.insert_local_constant(name.into(), literal);
+                Ok(())
+            }
+
+            PatternMode::Alternative(assigned) => {
+                match self.environment.scope.get(name) {
+                    // This variable was defined in the Initial multi-pattern
+                    Some(initial) if self.initial_pattern_vars.contains(name) => {
+                        assigned.push(name.into());
+                        let initial_typ = initial.type_.clone();
+                        unify(initial_typ, literal.type_())
+                    }
+
+                    // This variable was not defined in the Initial multi-pattern
+                    _ => Err(UnifyError::ExtraVarInAlternativePattern { name: name.into() }),
+                }
+            }
+        }
+    }
```

Comment:
We only ever insert strings so the vast majority of the code head is dead code and cannot be tested in any way. We can instead have a `insert_constant_string` function that does much less, and we can know that it works.